File: ChessLab/Variants/BarricadesFEN.swift
import ChessKit

/// Traduction d'une FEN de Barricades vers une FEN que ChessKit lit
/// CORRECTEMENT.
///
/// La FEN du moteur porte la lettre du mur
/// (``BarricadesConfiguration/wallLetter``), que ChessKit ne connaît pas.
///
/// ## Ce que ChessKit en fait AUJOURD'HUI — mesuré, pas supposé
///
/// Version 0.17.0, sondée avant d'écrire ce fichier : un caractère inconnu au
/// MILIEU d'une rangée fait avancer d'une colonne sans rien poser. Le mur
/// devient donc une case vide, et le reste de la rangée reste à sa place.
/// `rnbqkbnr/pppppppp/8/3pW1n1/…` se relit `…/3p2n1/…` : le pion d5 et le
/// cavalier g5 sont exactement où ils doivent être.
///
/// Autrement dit, ce filtre n'est pas ce qui sauve l'affichage aujourd'hui.
/// Il existe pour deux raisons quand même :
///
/// - **Ne pas dépendre d'une tolérance non écrite.** Rien, dans ChessKit, ne
///   promet ce comportement. Le Crazyhouse a déjà montré ce que coûte le pari
///   inverse : sa réserve, elle, arrive APRÈS la 8e colonne de la dernière
///   rangée, `Square.File` plafonne, et les caractères en trop s'empilent sur
///   h1 en masquant la tour (voir ``CrazyhouseFEN``).
/// - **Un seul chemin pour toutes les variantes.** ``VariantFEN`` compose les
///   deux nettoyages, si bien qu'aucun écran n'a à savoir laquelle des deux
///   enrichit sa FEN.
///
/// La règle reste celle du Crazyhouse : tout ce qui part vers ChessKit passe
/// par ``forChessKit(_:)`` ; tout ce qui part vers le MOTEUR garde la FEN
/// brute, seule à porter les murs.
enum BarricadesFEN {

    /// Remplace chaque mur par une case vide — le plateau seul, tel que
    /// ChessKit sait le lire.
    static func forChessKit(_ fen: String) -> String {
        var fields = fen.split(separator: " ", omittingEmptySubsequences: false).map(String.init)
        // Le test porte sur le PLACEMENT seul, jamais sur la FEN entière : le
        // champ du trait vaut « w » une fois sur deux, et le chercher partout
        // ferait retraiter chaque position de toutes les variantes pour rien.
        guard let placement = fields.first, placement.contains(where: isWall) else { return fen }

        fields[0] = placement.split(separator: "/", omittingEmptySubsequences: false)
            .map { compress(expand(String($0))) }
            .joined(separator: "/")
        return fields.joined(separator: " ")
    }

    /// Les cases murées d'une FEN. Lues plutôt que supposées : c'est le moteur
    /// qui fait autorité sur le plateau, ici comme ailleurs.
    static func wallSquares(in fen: String) -> [Square] {
        guard let placement = fen.split(separator: " ").first else { return [] }
        var squares: [Square] = []
        for (index, rankText) in placement.split(separator: "/", omittingEmptySubsequences: false).enumerated() {
            let rank = 8 - index
            guard (1...8).contains(rank) else { continue }
            let files = Array("abcdefgh")
            for (index, slot) in expand(String(rankText)).enumerated() where isWall(slot) {
                guard index < files.count else { continue }
                squares.append(Square("\(files[index])\(rank)"))
            }
        }
        return squares
    }

    // MARK: Rangée

    private static func isWall(_ character: Character) -> Bool {
        Character(character.uppercased()) == BarricadesConfiguration.wallLetter
    }

    /// Une rangée en 8 cases exactement, `.` pour une case vide.
    private static func expand(_ rank: String) -> [Character] {
        var slots: [Character] = []
        for character in rank {
            if let empty = character.wholeNumberValue, (1...8).contains(empty) {
                slots.append(contentsOf: Array(repeating: ".", count: empty))
            } else {
                slots.append(character)
            }
        }
        return slots
    }

    /// L'inverse, murs compris — ils comptent pour du vide.
    private static func compress(_ slots: [Character]) -> String {
        var result = ""
        var empty = 0
        for slot in slots {
            if slot == "." || isWall(slot) {
                empty += 1
            } else {
                if empty > 0 { result += "\(empty)"; empty = 0 }
                result.append(slot)
            }
        }
        if empty > 0 { result += "\(empty)" }
        return result
    }
}

File: android/app/src/main/java/com/chesslab/variants/BarricadesFen.kt
package com.chesslab.variants

import chesskit.Square

/**
 * La FEN des Barricades, et la protection de `chesskit` contre elle.
 *
 * Le mur est une PIÈCE aux yeux du moteur — la lettre « W », un type
 * `immobile` qui ne peut jouer aucun coup. `chesskit` ne la connaît pas et
 * refuserait la position entière à cause d'elle : on la lui retire avant de
 * lui donner quoi que ce soit, et l'on garde les cases de côté pour les
 * dessiner. Pendant de `BarricadesFEN.swift`.
 */
object BarricadesFen {

    /** La lettre du mur dans la FEN du moteur — MAJUSCULE, donc blanche. */
    const val wallLetter = 'W'

    private val files = "abcdefgh"

    /** La même FEN, sans les murs : celle que `chesskit` sait lire. */
    fun forChessKit(fen: String): String {
        val fields = fen.split(" ")
        val placement = fields.firstOrNull() ?: return fen
        // Le test porte sur le PLACEMENT seul, jamais sur la FEN entière : le
        // champ du trait vaut « w » une fois sur deux, et le chercher partout
        // ferait retraiter chaque position de toutes les variantes pour rien.
        if (placement.none { isWall(it) }) return fen
        val cleaned = placement.split("/").joinToString("/") { compress(expand(it).map { s -> if (isWall(s)) '.' else s }) }
        return (listOf(cleaned) + fields.drop(1)).joinToString(" ")
    }

    /**
     * Les cases murées d'une FEN. LUES plutôt que supposées : c'est le moteur
     * qui fait autorité sur le plateau, ici comme ailleurs.
     */
    fun wallSquares(fen: String): List<Square> {
        val placement = fen.split(" ").firstOrNull() ?: return emptyList()
        val squares = ArrayList<Square>()
        placement.split("/").forEachIndexed { index, rankText ->
            val rank = 8 - index
            if (rank !in 1..8) return@forEachIndexed
            expand(rankText).forEachIndexed { file, slot ->
                if (isWall(slot) && file < files.length) squares += Square("${files[file]}$rank")
            }
        }
        return squares
    }

    /** Réécrit une FEN SANS mur en y posant ceux qu'on lui donne. */
    fun inserting(walls: Set<Square>, fen: String): String? {
        val fields = fen.split(" ")
        val placement = fields.firstOrNull() ?: return null
        val ranks = ArrayList<String>()
        placement.split("/").forEachIndexed { index, rankText ->
            val rank = 8 - index
            val slots = expand(rankText).toMutableList()
            for (file in slots.indices) {
                if (slots[file] != '.' || file >= files.length) continue
                if (Square("${files[file]}$rank") in walls) slots[file] = wallLetter
            }
            ranks += compress(slots)
        }
        return (listOf(ranks.joinToString("/")) + fields.drop(1)).joinToString(" ")
    }

    private fun isWall(c: Char): Boolean = c.uppercaseChar() == wallLetter

    /** Une rangée de FEN en 8 cases, les vides en « . ». */
    private fun expand(rank: String): List<Char> {
        val slots = ArrayList<Char>(8)
        for (c in rank) {
            val empty = c.digitToIntOrNull()
            if (empty != null && empty in 1..8) repeat(empty) { slots += '.' } else slots += c
        }
        return slots
    }

    private fun compress(slots: List<Char>): String {
        val out = StringBuilder()
        var empty = 0
        for (slot in slots) {
            if (slot == '.') {
                empty++
            } else {
                if (empty > 0) { out.append(empty); empty = 0 }
                out.append(slot)
            }
        }
        if (empty > 0) out.append(empty)
        return out.toString()
    }
}
